File: research/arz/lootgraph.py
"""Reverse index: item record -> the enemies that can drop it.

The edges all live in database.arz:

    records/creatures/enemies/...   loot*Item*/loot*TableName* fields -> loot table records
    records/items/loottables/...    lootName*/lootTableName*/item*  -> item records or further tables

Loot tables nest (a creature's table points at other tables), so reaching the items means walking that
graph transitively. Weights (lootWeight*) are carried along but deliberately not turned into a drop
percentage here: the real chance also depends on the creature's own loot roll counts, difficulty scaling
and whether the table is a "dynamic" one that rolls a base item plus separate affixes. "Can drop" is
defensible from this data; "drops 0.4% of the time" is not.
"""

import sys
from collections import defaultdict

from arz import Arz

LOOT_KEY_HINTS = ("loot", "item")


def is_record(value):
    return isinstance(value, str) and value.endswith(".dbr")


def build(arz):
    """Returns (table_children, creature_tables) as record-path -> [record-path]."""
    table_children = {}
    creature_tables = defaultdict(list)

    for record in arz.records:
        name = record[0]
        is_table = name.startswith(("records/items/loottables/", "records/items/lootchests/"))
        is_creature = name.startswith("records/creatures/")

        if not (is_table or is_creature):
            continue

        try:
            fields = arz.fields(record)
        except Exception:
            continue

        refs = []
        for key, values in fields.items():
            if not any(hint in key.lower() for hint in LOOT_KEY_HINTS):
                continue
            refs.extend(v for v in values if is_record(v))

        if is_table:
            table_children[name] = refs
        elif refs:
            creature_tables[name] = refs

    return table_children, creature_tables


def sources_by_item(table_children, creature_tables):
    """item record -> set(creature records), following loot tables transitively."""
    item_sources = defaultdict(set)

    for creature, roots in creature_tables.items():
        seen, stack = set(), list(roots)
        while stack:
            node = stack.pop()
            if node in seen:
                continue
            seen.add(node)

            if node in table_children:
                stack.extend(table_children[node])
            elif node.startswith("records/items/"):
                item_sources[node].add(creature)

    return item_sources


def main():
    arz = Arz(sys.argv[1] if len(sys.argv) > 1 else "database.arz")
    table_children, creature_tables = build(arz)
    item_sources = sources_by_item(table_children, creature_tables)

    print(f"loot tables      : {len(table_children)}")
    print(f"creatures w/loot : {len(creature_tables)}")
    print(f"items with source: {len(item_sources)}")

    for probe in sys.argv[2:]:
        matches = [i for i in item_sources if probe in i]
        for item in matches[:3]:
            names = sorted(item_sources[item])
            print(f"\n{item}\n  {len(names)} source(s), e.g.:")
            for n in names[:6]:
                print(f"    {n}")


if __name__ == "__main__":
    main()

File: IAGrim/Parsers/Arz/LootGraph.cs
using IAGrim.Parser.Arz;
using log4net;

namespace IAGrim.Parsers.Arz {

    /// <summary>
    /// "Dropped by" / "Crafted from" for an item, derived from the game's own database.arz.
    ///
    /// Three record families carry the edges, and IAGD's normal parse throws two of them away (its
    /// ArzParser.IsInteresting explicitly excludes "/loottables/" and "/lootchests/"):
    ///
    ///   records/creatures/...      loot*Item*/loot*TableName* -> loot table records
    ///   records/items/loottables/  lootName*/lootTableName*   -> item records, or further tables
    ///   records/items/crafting/    the blueprint's output item
    ///
    /// Tables nest, so reaching the items means walking that graph transitively.
    ///
    /// WHAT THIS DELIBERATELY DOES NOT DO: turn lootWeight* into a drop percentage. The real chance also
    /// depends on the creature's loot roll count, difficulty scaling, and whether the table rolls a base item
    /// with separately-rolled affixes. "Can drop" is supported by this data; a percentage is not.
    ///
    /// Most gear has no named source at all - Grim Dawn rolls it from level-appropriate dynamic pools, which
    /// is why GrimTools shows "Random drop" for the bulk of the database. Roughly 1300 of 34000 records
    /// resolve to a specific creature.
    /// </summary>
    public class LootGraph {
        private static readonly ILog Logger = LogManager.GetLogger(typeof(LootGraph));

        private readonly Dictionary<string, List<string>> _sourcesByItem = new();
        private readonly Dictionary<string, List<string>> _blueprintsByItem = new();

        /// <summary>Record -> its description tag, for turning a source or blueprint into a display name.</summary>
        private readonly Dictionary<string, string> _creatureTags = new();

        public bool IsLoaded { get; private set; }

        public IReadOnlyList<string> SourcesFor(string? itemRecord) {
            if (string.IsNullOrEmpty(itemRecord) || !_sourcesByItem.TryGetValue(itemRecord!, out var sources)) {
                return Array.Empty<string>();
            }

            return sources;
        }

        public IReadOnlyList<string> BlueprintsFor(string? itemRecord) {
            if (string.IsNullOrEmpty(itemRecord) || !_blueprintsByItem.TryGetValue(itemRecord!, out var blueprints)) {
                return Array.Empty<string>();
            }

            return blueprints;
        }

        public string? TagFor(string creatureRecord) {
            return _creatureTags.TryGetValue(creatureRecord, out var tag) ? tag : null;
        }

        /// <summary>
        /// Reads database.arz and builds the reverse index. Takes a couple of seconds and allocates while it
        /// runs, so callers should do this off the UI thread; the maps are only published once complete.
        /// </summary>
        public void Load(string arzFile) {
            if (!File.Exists(arzFile)) {
                Logger.Warn($"No database.arz at {arzFile}; drop sources will be unavailable");
                return;
            }

            var started = System.Diagnostics.Stopwatch.StartNew();

            var tableChildren = new Dictionary<string, List<string>>();
            var creatureTables = new Dictionary<string, List<string>>();

            foreach (var record in ArzRecordReader.Read(arzFile)) {
                var isTable = record.Name.StartsWith("records/items/loottables/") ||
                              record.Name.StartsWith("records/items/lootchests/");
                var isCreature = record.Name.StartsWith("records/creatures/");
                var isBlueprint = record.Name.StartsWith("records/items/crafting/");

                if (!isTable && !isCreature && !isBlueprint) {
                    continue;
                }

                // A blueprint names its output in artifactName - no "loot" or "item" in the key, so it has to
                // be read explicitly rather than through the loot-field sweep below.
                if (isBlueprint) {
                    // Its own description tag, so the UI can print "Blueprint: Leviathan" rather than the
                    // record's file name.
                    if (record.Fields.TryGetValue("description", out var tag) && tag.Count > 0) {
                        _creatureTags[record.Name] = tag[0];
                    }

                    if (record.Fields.TryGetValue("artifactName", out var crafted)) {
                        foreach (var produced in crafted.Where(r => r.EndsWith(".dbr"))) {
                            Add(_blueprintsByItem, produced, record.Name);
                        }
                    }

                    continue;
                }

                var refs = new List<string>();
                foreach (var field in record.Fields) {
                    var key = field.Key.ToLowerInvariant();

                    if (isCreature && key == "description" && field.Value.Count > 0) {
                        _creatureTags[record.Name] = field.Value[0];
                        continue;
                    }

                    if (!key.Contains("loot") && !key.Contains("item")) {
                        continue;
                    }

                    refs.AddRange(field.Value.Where(v => v.EndsWith(".dbr")));
                }

                if (isTable) {
                    tableChildren[record.Name] = refs;
                }
                else if (refs.Count > 0) {
                    creatureTables[record.Name] = refs;
                }
            }

            foreach (var (creature, roots) in creatureTables) {
                foreach (var item in Reachable(roots, tableChildren)) {
                    Add(_sourcesByItem, item, creature);
                }
            }

            IsLoaded = true;
            Logger.Info($"Loot graph: {tableChildren.Count} tables, {creatureTables.Count} creatures, " +
                        $"{_sourcesByItem.Count} items with a named source, {_blueprintsByItem.Count} craftable " +
                        $"({started.ElapsedMilliseconds} ms)");
        }

        /// <summary>Walks the loot tables reachable from one creature's roots, yielding the item records.</summary>
        private static IEnumerable<string> Reachable(List<string> roots, Dictionary<string, List<string>> tableChildren) {
            var seen = new HashSet<string>();
            var stack = new Stack<string>(roots);

            while (stack.Count > 0) {
                var node = stack.Pop();
                if (!seen.Add(node)) {
                    continue;
                }

                if (tableChildren.TryGetValue(node, out var children)) {
                    foreach (var child in children) {
                        stack.Push(child);
                    }
                }
                else if (node.StartsWith("records/items/")) {
                    yield return node;
                }
            }
        }

        private static void Add(Dictionary<string, List<string>> map, string key, string value) {
            if (!map.TryGetValue(key, out var list)) {
                map[key] = list = new List<string>();
            }

            if (!list.Contains(value)) {
                list.Add(value);
            }
        }
    }
}
